merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java

```diff
@@ -214,62 +219,78 @@ private <Return> void stepTaskHelper(
-    } else if (status instanceof TaskStatus.CallingTask<Return> s) {
-      final var target = TaskId.generate();
-      SimulationEngine.this.tasks.put(target, new ExecutionState.InProgress<>(currentTime, s.child().create(this.executor)));
-      SimulationEngine.this.taskParent.put(target, task);
-      SimulationEngine.this.taskChildren.computeIfAbsent(task, $ -> new HashSet<>()).add(target);
-      frame.signal(JobId.forTask(target));
-
-      this.tasks.put(task, progress.continueWith(s.continuation()));
-
-      final var targetExecution = this.tasks.get(target);
-      if (targetExecution == null) {
-        // TODO: Log that we saw a task ID that doesn't exist. Try to make this as visible as possible to users.
-        // pass -- nonexistent tasks will never complete
-      } else if (targetExecution instanceof ExecutionState.Terminated) {
-        this.scheduledJobs.schedule(JobId.forTask(task), SubInstant.Tasks.at(currentTime));
-      } else {
-        this.waitingTasks.subscribeQuery(task, Set.of(SignalId.forTask(target)));
-      }
-    } else if (status instanceof TaskStatus.AwaitingCondition<Return> s) {
+    } else if (status instanceof TaskStatus.CallingTask<?, ?, Output> s) {
+      processCallStatus(task, progress, frame, currentTime, s);
+    } else if (status instanceof TaskStatus.AwaitingCondition<Output> s) {
       final var condition = ConditionId.generate();
       this.conditions.put(condition, s.condition());
       this.scheduledJobs.schedule(JobId.forCondition(condition), SubInstant.Conditions.at(currentTime));
 
-      this.tasks.put(task, progress.continueWith(s.continuation()));
+      this.tasks.put(task, progress.continueWith(s.continuation(), Unit.UNIT));
       this.waitingTasks.subscribeQuery(task, Set.of(SignalId.forCondition(condition)));
     } else {
       throw new IllegalArgumentException("Unknown subclass of %s: %s".formatted(TaskStatus.class, status));
     }
   }
 
+  // This helper binds Input and Midput; if it were inlined, the captured `?`s would not propagate correctly,
+  // causing type errors.
+  private <Input, Midput, Output>
+  void processCallStatus(
+      final TaskId task,
+      final ExecutionState.InProgress<?, Output> progress,
+      final TaskFrame<JobId> frame,
+      final Duration currentTime,
+      final TaskStatus.CallingTask<Input, Midput, Output> s
+  ) {
+    final var child = TaskId.generate();
+
+    // Waiting for Midput; will produce Output
+    final var blocked = progress.blockWith(s.continuation());
+    this.tasks.put(task, blocked);
+    this.waitingTasks.subscribeQuery(task, Set.of(SignalId.forTask(child)));
+
+    // Ready to run (with the given Input); will produce Midput
+    this.tasks.put(child, new ExecutionState.InProgress<>(
+        currentTime,
+        s.child().create(this.executor),
+        new MutableObject<>(s.input()),
+        blocked.input()::setValue));
```

Comment:
I keep looking for a cast somewhere - because when I prototyped this myself about a year ago, I needed to cast the child's return value to the parent's expectation. However, I think you've managed to use generics to guarantee that the types line up. Nice! 🚀